### PR TITLE
LF: update DAML-LF specification with exercise_by_key.

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
Here we add the specification for the new update introduced in #7663.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/spec/daml-lf-1.rst [hunk 1]
```diff
@@ -1248,6 +1251,14 @@ Then we define *well-formed expressions*. ::
     ——————————————————————————————————————————————————————————————— UpdExerciseWithouActors
       Γ  ⊢  'exercise_without_actors' @Mod:T Ch e₁ e₂  : 'Update' σ
 
+      'tpl' (x : T)
+          ↦ { …, 'choices' { …, 'choice' ChKind Ch (y : 'ContractId' Mod:T) (z : τ) : σ 'by' … ↦ …, …, 'key' τₖ …  } }
+        ∈ 〚Ξ〛Mod
+      Γ  ⊢  e₁  :  τₖ
+      Γ  ⊢  e₂  :  τ
+    ——————————————————————————————————————————————————————————————— UpdExerciseByKey
+      Γ  ⊢  'exercise' @Mod:T Ch e₁ e₂  : 'Update' σ
```
Comment (hurryabit) [inline daml-lf/spec/daml-lf-1.rst:1260]:
```suggestion
      Γ  ⊢  'exercise_by_key' @Mod:T Ch e₁ e₂  : 'Update' σ
```

File: daml-lf/spec/daml-lf-1.rst [hunk 2]
```diff
@@ -2889,6 +2920,17 @@ as described by the ledger model::
...
+     'tpl' (x : T) ↦ { …, 'key' @σ eₖ eₘ }  ∈ 〚Ξ〛Mod
+     'fetch_by_key' @Mod:T vₖ ‖ (st; keys)  ⇓ᵤ  Err t
+   —————————————————————————————————————————————————————————————————————— EvUpdExercByKeyFetchErr
+     'exercise_by_key' Mod:T.Ch cid vₖ v ‖ (st; keys)  ⇓ᵤ  Err t
+
+     'tpl' (x : T) ↦ { …, 'key' @σ eₖ eₘ }  ∈ 〚Ξ〛Mod
+     'lookup_by_key' @Mod:T vₖ ‖ (st; keys)  ⇓ᵤ  Ok ⟨'contractId': cid, 'contract': vₜ⟩ ‖ (st'; keys')
+     'exercise_without_actor' Mod:T.Ch v₁ v ‖ (st'; keys')  ⇓ᵤ  ur
+   —————————————————————————————————————————————————————————————————————— EvUpdExercByKeyExercise
+     'exercise_by_key' Mod:T.Ch cid vₖ v₁ ‖ (st; keys)  ⇓ᵤ  ur
```
Comment (hurryabit) [inline daml-lf/spec/daml-lf-1.rst:2932]: Something is off here but I can't exactly tell what. Should it be
```suggestion
     'exercise_without_actor' Mod:T.Ch cid v₁ ‖ (st'; keys')  ⇓ᵤ  ur
   —————————————————————————————————————————————————————————————————————— EvUpdExercByKeyExercise
     'exercise_by_key' Mod:T.Ch vₖ v₁ ‖ (st; keys)  ⇓ᵤ  ur
```

**Review by hurryabit — COMMENTED** [review]
Something is off about the operational semantics. I haven't fully figured out what yet...

File: daml-lf/spec/daml-lf-1.rst [hunk 3]
```diff
@@ -700,6 +702,7 @@ Then we can define our kinds, types, and expressions::
        |  'fetch' @Mod:T e                          -- UpdateFetch
        |  'exercise' @Mod:T Ch e₁ e₂ e₃             -- UpdateExercise
        |  'exercise_without_actors' @Mod:T Ch e₁ e₂ -- UpdateExerciseWithoutActors
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/spec/daml-lf-1.rst:704]: As exercise is supported only by LF < 1.5, we should just kill `exercise` and rename `exercise_without_actors` into exercise.

[the remark this answers is not in the document]
Comment (hurryabit) [reply daml-lf/spec/daml-lf-1.rst:704]: I'm fine with that too.

File: daml-lf/spec/daml-lf-1.rst [hunk 4]
```diff
@@ -2889,6 +2920,17 @@ as described by the ledger model::
        ⇓ᵤ
      Ok ('Some' @('ContractId' Mod:T) cid, ε) ‖ (st; keys)
 
+     'tpl' (x : T) ↦ { …, 'key' @σ eₖ eₘ }  ∈ 〚Ξ〛Mod
+     'fetch_by_key' @Mod:T vₖ ‖ (st; keys)  ⇓ᵤ  Err t
+   —————————————————————————————————————————————————————————————————————— EvUpdExercByKeyFetchErr
+     'exercise_by_key' Mod:T.Ch cid vₖ v ‖ (st; keys)  ⇓ᵤ  Err t
```
Comment (hurryabit) [inline daml-lf/spec/daml-lf-1.rst:2926]:
```suggestion
     'exercise_by_key' Mod:T.Ch vₖ v₁ ‖ (st; keys)  ⇓ᵤ  Err t
```

**Review by hurryabit — APPROVED** [review]
I think looks almost correct now. Thanks a lot!

File: daml-lf/spec/daml-lf-1.rst [hunk 5]
```diff
@@ -2889,6 +2920,17 @@ as described by the ledger model::
        ⇓ᵤ
      Ok ('Some' @('ContractId' Mod:T) cid, ε) ‖ (st; keys)
 
+     'tpl' (x : T) ↦ { …, 'key' @σ eₖ eₘ }  ∈ 〚Ξ〛Mod
+     'fetch_by_key' @Mod:T vₖ ‖ (st; keys)  ⇓ᵤ  Err t
+   —————————————————————————————————————————————————————————————————————— EvUpdExercByKeyFetchErr
+     'exercise_by_key' Mod:T.Ch vₖ v₁ ‖ (st; keys)  ⇓ᵤ  Err t
+
+     'tpl' (x : T) ↦ { …, 'key' @σ eₖ eₘ }  ∈ 〚Ξ〛Mod
+     'lookup_by_key' @Mod:T vₖ ‖ (st; keys)  ⇓ᵤ  Ok ⟨'contractId': cid, 'contract': vₜ⟩ ‖ (st'; keys')
```
Comment (sofiafaro-da) [inline daml-lf/spec/daml-lf-1.rst:2929]: You used `fetch_by_key` in one rule and `lookup_by_key` in the other.
I think it should be `fetch_by_key` in both cases?

```suggestion
     'fetch_by_key' @Mod:T vₖ ‖ (st; keys)  ⇓ᵤ  Err t
   —————————————————————————————————————————————————————————————————————— EvUpdExercByKeyFetchErr
     'exercise_by_key' Mod:T.Ch vₖ v₁ ‖ (st; keys)  ⇓ᵤ  Err t

     'tpl' (x : T) ↦ { …, 'key' @σ eₖ eₘ }  ∈ 〚Ξ〛Mod
     'fetch_by_key' @Mod:T vₖ ‖ (st; keys)  ⇓ᵤ  Ok ⟨'contractId': cid, 'contract': vₜ⟩ ‖ (st'; keys')
```

File: daml-lf/spec/daml-lf-1.rst [hunk 6]
```diff
@@ -2835,7 +2867,6 @@ as described by the ledger model::
 
      'tpl' (x : T) ↦ { …, 'key' @σ eₖ eₘ }  ∈ 〚Ξ〛Mod
      (eₘ vₖ)  ⇓  Err t
-     (Mod:T, vₖ) ∉ dom(keys₀)
```
Comment (sofiafaro-da) [inline daml-lf/spec/daml-lf-1.rst:2838]: good catch!

File: daml-lf/spec/daml-lf-1.rst [hunk 7]
```diff
@@ -2889,6 +2920,17 @@ as described by the ledger model::
...
+     'tpl' (x : T) ↦ { …, 'key' @σ eₖ eₘ }  ∈ 〚Ξ〛Mod
+     'fetch_by_key' @Mod:T vₖ ‖ (st; keys)  ⇓ᵤ  Err t
+   —————————————————————————————————————————————————————————————————————— EvUpdExercByKeyFetchErr
+     'exercise_by_key' Mod:T.Ch vₖ v₁ ‖ (st; keys)  ⇓ᵤ  Err t
+
+     'tpl' (x : T) ↦ { …, 'key' @σ eₖ eₘ }  ∈ 〚Ξ〛Mod
+     'lookup_by_key' @Mod:T vₖ ‖ (st; keys)  ⇓ᵤ  Ok ⟨'contractId': cid, 'contract': vₜ⟩ ‖ (st'; keys')
+     'exercise_without_actor' Mod:T.Ch cid v₁ ‖ (st'; keys')  ⇓ᵤ  ur
+   —————————————————————————————————————————————————————————————————————— EvUpdExercByKeyExercise
+     'exercise_by_key' Mod:T.Ch vₖ v₁ ‖ (st; keys)  ⇓ᵤ  ur
```
Comment (sofiafaro-da) [inline daml-lf/spec/daml-lf-1.rst:2932]: This is a nice approach, but the `ur` includes the ledger transaction, which means we aren't distinguishing between "exercise" and "exercise by key". Maybe the semantics needs to be reformulated anyway to track the transaction structure more closely, so I'm ok leaving it like this for now.

File: daml-lf/spec/daml-lf-1.rst (hunk 5 above)
Comment (remyhaemmerle-da) [reply daml-lf/spec/daml-lf-1.rst]: Good catch.

File: daml-lf/spec/daml-lf-1.rst (hunk 7 above)
Comment (remyhaemmerle-da) [reply daml-lf/spec/daml-lf-1.rst:2932]: You are right, this does not make the distinction between "exercise" and "exercise by key". But currently the implementation does not neither, so it is probably fine to let the semantics like that until we introduce a new transaction format.

Comment (hurryabit) [reply daml-lf/spec/daml-lf-1.rst:2932]: Let's leave it as is for now and fix it when we've introduced the new transaction format.

**Review by sofiafaro-da — APPROVED** [review]
LGTM, thank you!

**Pull request merged**